ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/StreamSelectBuilder.java

```diff
@@ -38,28 +45,74 @@ private StreamSelectBuilder() {
     final QueryContext queryContext = step.getProperties().getQueryContext();
 
     final LogicalSchema sourceSchema = stream.getSchema();
+    final Optional<ImmutableList<ColumnName>> selectedKeys = step.getSelectedKeys();
 
     final Selection<K> selection = Selection.of(
         sourceSchema,
         step.getKeyColumnNames(),
+        selectedKeys,
         step.getSelectExpressions(),
         buildContext.getKsqlConfig(),
         buildContext.getFunctionRegistry()
     );
 
+    final ImmutableList.Builder<Integer> keyIndexBuilder = ImmutableList.builder();
+    if (selectedKeys.isPresent()) {
+      final ImmutableList<ColumnName> keyNames = sourceSchema.key().stream()
+          .map(Column::name)
+          .collect(ImmutableList.toImmutableList());
+
+      for (final ColumnName keyName : selectedKeys.get()) {
+        keyIndexBuilder.add(keyNames.indexOf(keyName));
+      }
+    }
+    final ImmutableList<Integer> keyIndices = keyIndexBuilder.build();
+
     final SelectValueMapper<K> selectMapper = selection.getMapper();
 
     final ProcessingLogger logger = buildContext.getProcessingLogger(queryContext);
 
     final Named selectName =
         Named.as(StreamsUtil.buildOpName(queryContext));
 
-    return stream.withStream(
-        stream.getStream().transformValues(
-            () -> new KsTransformer<>(selectMapper.getTransformer(logger)),
+    if (selectedKeys.isPresent() && !selectedKeys.get().containsAll(
+        sourceSchema.key().stream().map(Column::name).collect(ImmutableList.toImmutableList())
+    )) {
+      return stream.withStream(
+          stream.getStream().transform(
+            () -> new KsTransformer<>(
```

Comment:
`KsTransformer` implements `Transformer` interface and is allowed to change the key. `KsValueTransformer` implement `ValueTransformer` interface and is not allowed to changed the key.

If the key is (potentially) changed, it might trigger a data repartition step downstream. Thus, we prefer non-key-changing step strictly over key-changing ones, expect, we actually really need to change the key.